### PR TITLE
SI-8154 AnyRefMap iterates its way to ((null, null))

### PR DESCRIPTION
**Pull request by Ichoran**
Changed logic to prevent mutation between hasNext and next from delivering invalid results.

### CODE REVIEW

**Comment by som-snytt** [comment]
I added a typo and format fix for the class scaladoc, but for some reason I can't PR to your branch.  It's just a couple of chars for the `^superscript^`:

https://github.com/som-snytt/scala/compare/review;8154

File: src/library/scala/collection/mutable/AnyRefMap.scala [hunk 1]
```diff
@@ -22,9 +22,9 @@ import generic.CanBuildFrom
  *  on a map that will no longer have elements removed but will be
  *  used heavily may save both time and storage space.
  * 
- *  This map is not indended to contain more than 2^29 entries (approximately
- *  500 million).  The maximum capacity is 2^30, but performance will degrade
- *  rapidly as 2^30 is approached.
+ *  This map is not indended to contain more than 2^29^ entries (approximately
```
Comment (gkossakowski) [inline src/library/scala/collection/mutable/AnyRefMap.scala]: s/indended/intended

**Comment by adriaanm** [comment]
review by @gkossakowski please

**Comment by gkossakowski** [comment]
@Ichoran: this PR can't be merged anymore. Please rebase it on top of master and fix the typo in your doc change.

**Comment by Ichoran** [comment]
@gkossakowski - Yeah, I know, it's on my to-do list.  My own commits did it in.  I'm too busy through Tuesday to get to it, though.

**Comment by Ichoran** [comment]
@gkossakowski - Ready for review, tests finally pass.  It's a little ugly code-duplication-wise.

File: src/library/scala/collection/mutable/AnyRefMap.scala [hunk 2]
```diff
@@ -22,9 +22,9 @@ import generic.CanBuildFrom
  *  on a map that will no longer have elements removed but will be
  *  used heavily may save both time and storage space.
  * 
- *  This map is not indended to contain more than 2^29 entries (approximately
- *  500 million).  The maximum capacity is 2^30, but performance will degrade
- *  rapidly as 2^30 is approached.
+ *  This map is not intended to contain more than 2^29^ entries (approximately
+ *  500 million).  The maximum capacity is 2^30^, but performance will degrade
+ *  rapidly as 2^30^ is approached.
```
Comment (gkossakowski) [inline src/library/scala/collection/mutable/AnyRefMap.scala:27]: typo in "2^30^"

Comment (Ichoran) [inline src/library/scala/collection/mutable/AnyRefMap.scala:27]: What typo?  I saw "typo" and fixed it in LongMap which was actually broken.  But isn't N^M^ valid superscript?

Comment (gkossakowski) [inline src/library/scala/collection/mutable/AnyRefMap.scala:27]: Oh, I totally misunderstood your change. Sorry.

**Comment by gkossakowski** [comment]
LGTM

Thanks!

**Pull request merged**